### DISCUSSION TITLE
Seeking Advice: GitHub SSO User Access in Teleport without Shared Database Accounts

### DISCUSSION THREAD

**Discussion by computerphilosopher** [gravitational/teleport#30529] [Q&A] [answered]
Hello everyone!

I'm currently leveraging the open-source edition of Teleport for managing database access control. My goal is to transition the authentication method from local to GitHub SSO.

After adding the database account to the db_users traits for a user set up through GitHub authentication, I noticed that upon re-login, the values reverted to their defaults.

Is there a way to grant GitHub SSO users access to the database account without allowing sharing of the account among users within the same team? I'd like each user to use their own unique database account.

**Comment by webvictim** [comment, marked as answer]
There's actually a hidden little trick with Github where your Github username (`logins`) and Github teams (`github_teams`) are automatically exposed under the `{{internal.<x>}}` bag.  You can see the traits at the command line if you run `tsh -d status`:

```
gus@apollo:~ % tsh -d status
...
> Profile URL:        https://teleport.example.com:443
  Logged in as:       webvictim
  Cluster:            purple
  Roles:              access
  Traits:             kubernetes_groups: []
                      kubernetes_users: []
                      logins: [webvictim]
                      github_teams: [admins]
  Logins:             webvictim, ubuntu, ec2-user
  Kubernetes:         enabled
  Valid until:        2023-08-17 01:26:19 -0300 ADT [valid for 12h0m0s]
  Extensions:         login-ip, permit-agent-forwarding, permit-port-forwarding, permit-pty, private-key-policy, teleport-device-asset-tag, teleport-device-credential-id, teleport-device-id
```

What this means is that you can reference the `{{internal.logins}}` trait under `db_users` in a Teleport role like this:

```yaml
spec:
  allow:
    db_users:
    - '{{internal.db_users}}'
    - '{{internal.logins}}'
```

Then when a Github user logs in, their certificate will have their Github username (`webvictim` for me) as an allowed database user:

```json
{
  "cert_type": "user",
  "cluster_name": "teleport.example.com",
  "code": "TC000I",
  "ei": 0,
  "event": "cert.create",
  "identity": {
    "client_ip": "1.2.3.4",
    "database_users": [
      "webvictim"
    ],
...
```

This would enable you to create a database user for every Github user which matches their Github login, if you wanted.

**Reply by computerphilosopher** [reply]
Thanks to your answer, I've solved the problem, but what you said doesn't seem to be easy to guess from the official documentation.

**Reply by webvictim** [reply]
I agree. I will take a look to see where I can add this information into the docs!
